### DISCUSSION TITLE
Does GPL cover the documents exported using Quarto when Quarto copies parts of itself into the documents?

### DISCUSSION THREAD

**Discussion by okayucob** [quarto-dev/quarto-cli#4639] [Q&A]
Thank you for developing Quarto. I’m trying creating proprietary html documents for my website using Quarto. The outputs are excellent, but I’m not sure that GPL does not cover the exported html documents when they include styles or functionalities from Quarto.

I checked the FAQ (https://quarto.org/docs/faq/#can-i-use-quarto-to-develop-proprietary-content) and understand the copyright on Quarto does not intend to cover the generated contents.

Let me explain the situation. I’m using this kind of settings in my _quarto.yml.

```
format:
  html:
    embed-resources: false
    code-copy: true
    highlight-style: arrow
    html-math-method: katex
    theme: none
    minimal: true
```

Therefore, Quarto will copy some CSS codes for syntax highlighting and JavaScript codes for copying codes in the documents from its source files into my documents.

As far as I understand, the copied source codes will inherit the original license (https://www.gnu.org/licenses/gpl-faq.en.html#CanIUseGPLToolsForNF). So the GPL will cover parts of my documents. If so, the document authors like me will need some work to be compliant with the license.

I think the license of Quarto does not intend to cover the exported documents, even if Quarto copies its source codes to give the documents some styles and functionalities. If so, as stated here (https://github.com/quarto-dev/quarto-cli/discussions/2915), distinguishing between the Quarto tools and “redistributable files” (CSS and JavaScript codes in my case?) may solve this question and similar ones.

Could someone clarify this?

Thank you for all of your work.

**Comment by mcanouil** [comment]
Source code is under GPL, but source code (SCSS, CSS, etc.) is not the content, which can be under a proprietary license.

Otherwise and for example, everything you write using Word is under Microsoft licensing, thus you don't own things you create with any tools.

**Reply by okayucob** [reply]
Thank you for the reply. Yes, I agree with you if we consider the “content” as something like “A document that a user has the copyright and created with Quarto“.

But the problem is when the exported document file has both the “content” and the copies of the source code. For example, when adding copy buttons in the html file with `code-copy: true`, Quarto seems to copy some of its source code below into the document:

- https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/html/templates/quarto-html.ejs

I think exporting html files containing this kind of source code is a normal use case for Quarto. We can make the “content” be under a non-GPL license, but the source code in the exported document file is not because it is licensed under GPL. It seems that the copyright on Quarto does not intend to cover the exported files (https://quarto.org/docs/faq/#can-i-use-quarto-to-develop-proprietary-content). So I’m not sure how the license of Quarto handles the document files containing both the content and the source code.

**Reply by mcanouil** [reply]
Same applies, a Word document contains Microsoft source code (xml files, etc.), it's still not content.

**Reply by okayucob** [reply]
Hi,

Thanks. But my question is about the GPL license and not about the license of documents generated with Microsoft Word. I usually don’t use the software. So I don’t know its license. However, a licensor of a software will decide whether there are restrictions to the output that contain such an embedded source code.

As for softwares licensed under GPL, copied source code in the output is, as far as I understand, still covered by GPL. You can check this here (https://www.gnu.org/licenses/gpl-faq.en.html#CanIUseGPLToolsForNF). I’m not sure how Quarto handles the copied source code with regard to GPL.

**Reply by mcanouil** [reply]
The code copied/embedded is under GPL but still has nothing to do with your content which can be proprietary.
You have to understand that these are two entirely different things.

I tried to explain this to you with my Word/OpenOffice example, they both embed code/etc. in addition to the content you might have written in it.

- If you intend to write documents/slides/websites with Quarto, everything you write is not restricted to Quarto licensing.
- If you intend to reuse Quarto as a tool or reuse Quarto code as part of a new software, you cannot unless it is also under the same license.

**Reply by okayucob** [reply]
Thank you for the reply and for taking your time.

As I wrote previously, I understand the difference between the content I created and the copied/embedded source code. I can make the content be under a non-GPL license even if the copied source code is under GPL.

Can I ask you two questions related to this? These will enhance my understanding of this.

Q.1
I might misunderstand the answer to the FAQ below.
In the FAQ (https://quarto.org/docs/faq/#can-i-use-quarto-to-develop-proprietary-content), there is a question:

“Can I use Quarto to develop proprietary content?”

In the above link, I’m not sure what “content“ really means. When I export html file with Quarto, do you think “content“ is the whole file, or it means the content that I have the copyright and it does not include the embedded source code?

Q.2
Do you think exporting html files with Quarto and publishing them in the scenario below requires me to follow the GPL?

I intend to write paid webpages/html files with Quarto. The pages will be behind a paywall. Quarto will copy some of its source code (CSS and JavaScript) into the html files. For now, I’m going to:
- put the exported proprietary contents behind the paywall and
- add the copied source code to the header so that I can add, for example, copy to clipboard buttons.

I think this is a redistribution of the source code so I will, for example, put a license notice on the webpages.

Thank you.

**Reply by mcanouil** [reply]
Q.1: content is not the code, content is text, figures, etc.

Q.2: No. You are not selling Quarto or a tool based on Quarto. You are selling access to content (created/published by Quarto as a tool).

You can acknowledge that HTML pages were created by Quarto, but this does not change the license of your content.

**Reply by okayucob** [reply]
Hi,

Thank you for the reply.

Q.1
I thought that the "content" included also the exported source code.
But now, I feel thinking as your answer is more proper. Thanks.

Q.2
However, I'm still confused. In the situation written in Q.2, it's true that I'm not selling Quarto but selling the access to my content. But at the same time, the visitors will access the exported source code licensed under GPL. I think it's a redistribution of the source code.

Although my understanding may not be correct, I will write some codes by myself to add styles and the functionalities to my contents this time.

**Reply by mcanouil** [reply]
With your understanding this would mean you are never allowed to create content and/or sell that content or access to it as long as you are using a tool not build by yourself.

See it as an painter making a piece of art. The painter can sell his painting, thus the content itself and the canvas on which lies the content. It does not matter how or by whom the canvas was made.


Quarto HTML documents include Quarto code, that's a fact.
As many (if not all) "creative" source code is included.
To be on the safe side, since you are not building a software, nor selling a software made upon Quarto, you can simply add an attribution statement: "This page was built using Quarto, an open source software released under GPL, and might include source code within the HTML code" or something similar.
This way, you clearly give attribution for "redistribution" but also separate your content from the software you used to display your content.

**Reply by okayucob** [reply]
> With your understanding this would mean you are never allowed to create content and/or sell that content or access to it as long as you are using a tool not build by yourself.

Hmm... I don't think so. If a software doesn't copy its ("creative") source code into the exported documents, images, files, etc, then there are no restrictions to the exported objects, I think.

Can I ask your thoughts about the case of Bison mentioned [here](https://www.gnu.org/licenses/gpl-faq.en.html#CanIUseGPLToolsForNF)?
I think the two cases are similar.

As for adding an attribution statement, yes, this would be another solution.

**Reply by mcanouil** [reply]
Well Word/Excel copy xml code in your documents.

The case mentioned is exactly what I am trying to tell you.

Really, simply add a statement and you'll be under the "fair use" case which is basically what Quarto FAQ sis stating.

**Reply by chriskmanx** [reply]
Hello Quarto Creators, I wanted to "briefly" chime in on this discussion. I apologize in advance for the double posting, but I intend to also make this comment  [here](https://github.com/quarto-dev/quarto-cli/discussions/2915) as that thread seems like an authoritative discussion. 

I spent some 10 years as an IT manager of a global hedge fund and licensing would of course, as for anyone in that role, come up where commercial software is involved. I am really quite taken by Quarto and would like to believe it will be possible to use it commercially given its GPL heritage, but I would like some clarification.  

I like the analogy by @mcanouil in respect of the Word Processor. Indeed, a strict interpretation of the line of thought by @okayucob would lead one to conclude that if I write a poem in AbiWord ( a Linux word processor distributed under the GPL ) that then the stylistic elements applied to that poem by AbiWord upon rendering the document, would prompt my poem also to be covered by the GPL --- if distributed. That's at the core of what people are concerned about here if I am not mistaken. 

Now, the GPL is intended to cover **software** and there are as of late many grey areas in the application of the GPL resulting from the evolution taking place in that space. Take the discussion around AI,  Chat-GPT, and LLaMa by MetaAI involving model weights which are not the output of a process of creative authorship covered by copyright law in the first instance. Indeed, there is a lawsuit underway in respect of Stable Diffusion, another AI package, as to whether the model weights within that software violate the copyright of the original content used in training the Stable Diffusion model. Who is right? 

Besides the editor/word processor analogy, there is perhaps another analogy that might apply here, given that Quarto is essentially a LaTeX processor or a derivative of a LaTeX processor with markdown processing added in. There are some interesting discussions on the internet, which essentially came to very different conclusions than the ones purported here. 

https://tex.stackexchange.com/questions/69007/the-gpl-and-latex-packages

The above concludes: "All cases must be licensed under the GPL, as they are all [modified versions](http://www.gnu.org/licenses/gpl-faq.html#MereAggregation), not aggregates. "

Also: 

https://opensource.stackexchange.com/questions/2735/gpl-licensed-latex-template-implications-for-resulting-work


The gentleman there has a very similar problem: combining a LaTeX template (GPL) with his proprietary work ( a thesis ) which is rendered using LaTeX. That discussion thread seems to firstly conclude that "Any derived work has to be licensed as GPL. GPL forces to make the source available. Thus, the tex source of the generated PDFs have to be made available, too" but then differentiates based on the fact that the thesis, as in my example of the poem, is data rather than code and that therefore the generated PDFs would indeed not be covered by the GPL. Now, in the case of Quarto, the inputs are code & data. Commercial authors are likely to want to protect both. This is incredibly important since if the Quarto system may be thought of as a compiler and functionalities & styles may be thought of as libraries, then any code input to that system would potentially be covered by the GPL. The word processor analogy would no longer apply. 

An argument could then be made that the code processed by Quarto is loosely coupled as it is being interpreted ( compiled )  by a forked process ( R, Python, Julia ) and that this loose coupling absolves the code from any obligations under the GPL. In reading the GPL, this only holds true if no "complex" data structures are being shared. 

This would be my first question to the creators of Quarto. Do you hold that the above argument ( loose coupling & no complex shared data structures ) holds true? 

My second question/request is that a license, GPL, or otherwise is an agreement between a content creator and a licensee. The GPL isn't a law in and of itself. Hence, the only party who could hold a licensee accountable is the content creator, not anyone else. Nobody else has recourse: Legal Framework for Open Source Software (in German) http://www.beck-shop.de/Jaeger-Metzger-Open-Source-
Software/productview.aspx?product=14503508 . Differently put, your ( Quarto creators ) stated position, therefore, is of paramount importance, as it relates to the "whole understanding" of the licensee entering a license agreement. In areas, where you indemnify your licensees, the GPL loses its strength. That would really be my second question. Are you prepared to indemnify your licensees in accordance with your stated intent, i.e. that intent being that original contenrt is not covered? 

None of the opinions stated above constitute legal advice. 

Best Regards

Chris Kohlhepp
